flocker/node/agents/test/test_blockdevice.py

```diff
@@ -5249,3 +5257,231 @@ def test_get_device_path_until_detach(self):
 
         self.assertRaises(UnattachedVolume,
                           self.cache.get_device_path, attached_id1)
+
+
+@attributes(["_proxy_object", "_callback"], apply_with_init=False)
+class CallbackProxy(object):
+    """
+    A proxy to another object that calls a callback before and after each call
+    to the underlying object. This enables tests where you want to inject code
+    before or after any calls to an underlying API.
+
+    Note that this proxy assumes that the underlying object only has public
+    methods and not public attributes.
```

Comment:
My inclination would be to require an interface here and only proxy the methods of that interface. ``flocker.common.interface_decorator`` would be useful here.